### ISSUE TITLE
Orders sometimes stuck with processing status "running"

### ISSUE THREAD

**Issue by loswochos** [PAYONE-GmbH/magento-1#119]
Since I patched our Magento 1.9.3.1 with SUPEE-10266 we have a strange issue with Payone. Shop orders (we are using paypal and credit card) stuck with status "running" in the transaction log (and "pending" in magento). This happens only sometimes and very unspecific (maybe 1-2 times a week). I can't find any connection between these orders.

At the moment I have to manually set the processing_status from "running" to "pennding" in the payone_protocol_transactionstatus table and the cron proccesses the order correctly during the next run.

Crontab:

*/5 *   * * *   /var/www/htdocs/magento/cron.sh

**Comment by rofthedeep** [comment]
Hi, we are getting similar problems since we updated our payone extension (Release v4.0.1) on monday. At the moment we are running Magento Version 1.9.3.1. We didn't install SUPEE-10266 yet.

We are getting these errors: Call to a member function hasForcedState() on boolean in  /app/code/core/Mage/Sales/Model/Order/Invoice.php on line 415

At the moment the problems occur with paypal and creditcard payments. Not every time, only sometimes.

In the payone transaction table we can see, that the payone payment_id can not be found. So, no payment object is loaded and the invoice can not be created. As a consequence the status of the order doesn't change correctly.

![bildschirmfoto 2017-11-10 um 12 39 31](https://user-images.githubusercontent.com/2420176/32656975-45ec8782-c614-11e7-9f1c-04fbb818ea0e.png)

A second error we found is the following:
![bildschirmfoto 2017-11-10 um 12 33 19](https://user-images.githubusercontent.com/2420176/32656720-604580b2-c613-11e7-995e-3f1b21a62ed8.png)

The capture of credit card payment seems to fail. Perhaps because of the missing subacccount id in the request?

Our workaround ist the same @loswochos :)

**Comment by loswochos** [comment]
@rofthedeep 

I've checked the magento system.log and apache/php error log and can't find an error like yours. But I can confirm the entries in the payone_protocol_transactionstatus table:

`Payment method configuration with id "14" not found.`
`Payment method configuration with id "15" not found.`

So it seems to be the same problem.

**Comment by rofthedeep** [comment]
@loswochos 
This commit seems to cause the problem:
https://github.com/PAYONE-GmbH/magento-1/commit/c4a433197428f563ad9d4ee1879efc3065281401#diff-b3fbdec37d55ddbe95e434e1d88e0c5a

With setting the store id to 0 we don't get the problem in our tests ...

**Comment by fjbender** [comment]
Guys, thanks for your feedback. As you can see from e.g. #35 we do seem to have a problem with orders being weird "sometimes" and haven't found a way of reproducing it systematically yet.

@rofthedeep 
With c4a433197428f563ad9d4ee1879efc3065281401 we tried to fix a bug where a wrong config was loaded from the cache when there were multiple payone configurations for different stores. However, I'm not quite sure what to do here, as this commit fixed the issue for the merchant but seemed to cause a new issue for you.

**Comment by rofthedeep** [comment]
@fjbender no problem :)

In our stores the payment_id switched between the values for the website and the values from the store (for example Credit Card):
![bildschirmfoto 2017-11-13 um 14 01 51](https://user-images.githubusercontent.com/2420176/32726986-4026e5fe-c87b-11e7-895d-19669b251a05.png)

The order has the payment id 45, the object loaded the payments based ob the website - so the object contains only the payment method with id 44.

**Comment by fjbender** [comment]
For the time being, I'll have to bet on this issue being "accidentally" fixed when we fix #35 and/or #8. I don't have any angles to work on this issue that aren't discussed in the other issues, I'm afraid.

**Comment by loswochos** [comment]
@fjbender 

Do you have a timeline until when the issues will be resolved?

**Comment by fjbender** [comment]
@loswochos Unfortunately, I don't. @tniebergall is working on the other issues as well, he'll post an update as soon as he finds out what this all is about.

**Comment by tniebergall** [comment]
@loswochos @rofthedeep could you send me a dump of the payone_config_payment_method, payone_protocol_api table and some orderIds of affected orders?

**Comment by Backenkoehler** [comment]
We have the same problem since we upgraded to 1.9.3.6 in October. Right now we change the status from running to pending and in the next batch / cronjob run the orders are getting processed. Seems like a bit odd.

Any suggestions what we can do here?

**Comment by fjbender** [comment]
@Backenkoehler unfortunately, this issue is currently somewhat of a mystery and we have no workaround other than manually "freeing" the order.

**Comment by Backenkoehler** [comment]
We have still daily 30 orders which are stuck. Can we expect any help for this?

**Comment by fjbender** [comment]
@tniebergall Anything new here?

**Comment by Backenkoehler** [comment]
Hey @fjbender,
Hey @tniebergall,

maybe we found something. Everytime an order gets stuck we have an Deadlock. Could we improve error handling to set this failed orders back to pending? Doing by hand with a sql query does the work right now.

Lines marked with ** throw the exception. Stacktrace:

PDOException: SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction in htdocs/lib/Zend/Db/Statement/Pdo.php:228

`Payone/Core/Model/Service/TransactionStatus/Execute.php in execute at line 95

        // Load Area to get Translation

        $this->getApp()->loadAreaPart(Mage_Core_Model_App_Area::AREA_FRONTEND, 
Mage_Core_Model_App_Area::PART_TRANSLATE);

        $this->getApp()->loadAreaPart(Mage_Core_Model_App_Area::AREA_FRONTEND, Mage_Core_Model_App_Area::PART_DESIGN);

        $transactionStatus->setStatusRunning();

        **$transactionStatus->save();**

        $this->helper()->logCronjobMessage("ID: {$transactionStatus->getId()} - Execute - Set status to running", $storeId);

        try {

            $this->getServiceProcess()->execute($transactionStatus);

            $transactionStatus->setStatusOk();

            $this->helper()->logCronjobMessage("ID: {$transactionStatus->getId()} - Execute - Finished service execution, set status to complete", $storeId);

`

`Payone/Core/Model/Service/TransactionStatus/Execute.php in executePending at line 63

            // Get next pending TransactionStatus

            $transactionStatus = $collection->getNextPending();

            // Execute

            if ($transactionStatus) {

                **$this->execute($transactionStatus);**

                $countExecuted++;

            }

            else {

                $continue = false;

            }`

`Payone/Core/Model/Cronjob/TransactionStatus/Worker.php in execute at line 49

    {

        Mage::helper('payone_core')->logCronjobMessage("Started cronjob");

        $service = $this->getFactory()->getServiceTransactionStatusExecute();

        **$count = $service->executePending();**

        $schedule->setMessages($count . ' Transaction-Status processed');

        Mage::helper('payone_core')->logCronjobMessage("Finished cronjob");

    }
`

**Comment by fjbender** [comment]
That's an interesting insight! I don't know if people still use it, but maybe https://github.com/AOEpeople/Aoe_DbRetry is an option to prelimiarily fix the issue?

**Comment by Backenkoehler** [comment]
We already use this extension and we have still deadlocks. It's a bit worrying.

**Comment by Backenkoehler** [comment]
Any Updates? We still have errors like this + orders stuck in pending. We need a solution or we have to evaluate other payment options. It's not acceptable to run custom queries on the live database daily if you do 6 figures through credit card.
> Payment method configuration with id "2" not found.

**Comment by fjbender** [comment]
@tniebergall Let‘s discuss this asap

**Comment by loswochos** [comment]
Same here. Problem still exists and we need a solution.

**Comment by Backenkoehler** [comment]
@fjbender Problem is still there. We have a hack for this, running some SQL but it's not acceptable to be honest.
`UPDATE payone_protocol_transactionstatus pa SET pa.processing_status = 'pending' WHERE pa.reference IN (SELECT * FROM (SELECT DISTINCT p.reference FROM payone_protocol_transactionstatus p LEFT JOIN sales_flat_order o ON p.order_id = o.entity_id LEFT JOIN sales_flat_invoice i ON i.order_id = p.order_id  WHERE p.processing_status = 'running' AND i.increment_id IS NULL AND p.created_at >= '2018-03-01 00:00:00' AND processed_at = '0000-00-00 00:00:00' AND o.status = 'processing' ORDER BY p.reference) temp_table); `

Do you have any solution? Should we contact someone of you on a different channel?

**Comment by fjbender** [comment]
@Backenkoehler Sure, let's have a talk about this. Drop me a line at integrations@bspayone.com and I'll get back to you. :)

**Comment by zaietsv** [comment]
I have some ideas about the cache behavior. I have created an extension which fixes this issue, as I hope. Now it requires some tests at a live environment.

**Comment by zaietsv** [comment]
There is a method Payone_Core_Model_Factory->getConfigProxyMode() which always calls a null-store configuration which falls back to a website configuration. If a configuration cache is clear at that moment, it caches this website configuration under a cache key "store_1_payone_config_cache". Every next method wich calls a "1-store" configuration obtains a cached "null-store" configuration since the first store cache key has the same name "store_1_payone_config_cache". It is not completely wrong, since most of fields has the same values in differnt configurations. 
But concerning to a payment method id value, it is different for the website and for the store. Every method calling a payment-id for a particular store obtains a different Id for website. At that moment appears an error message "Error during service execution, set status to error with message Payment method configuration with id "99" not found." . The script stops an invoice generation because it unable to extract a properly payment method Id.
How could you fix it? You need to create an unique cache-key name for null-store calls (i.e. website-config calls). Let it be "store_NULL_payone_config_cache" or "store_website_payone_config_cache" . Therefore all you need is to rewrite Payone_Core_Model_Service_InitializeConfig->getConfigCacheId() and Payone_Core_Model_Service_InitializeConfig->getConfigRegistryKey().
`<?php
/**
 * Created by IntelliJ IDEA.
 * User: vza
 * Date: 18.08.18
 * Time: 0:42
 */

class Lindenvalley_Payone_Core_Model_Service_InitializeConfig extends Payone_Core_Model_Service_InitializeConfig
{
    const NULL = 'NULL';

    /**
     * Returns a separate registry key for a "null" store config.
     * A parent method could occasionally override cached store configuration with a parent website configuration.
     * Usually it occurs under a registry key 'payone_core_config_1', where 1 is default store id.
     * @param int|null $storeId
     * @return string
     */
    public function getConfigRegistryKey($storeId = null)
    {
        if ($storeId === null) {
            $storeId = $this->getStoreId();
        }

        if ($storeId === null) {
            $storeId = self::NULL;
        }

        $cacheId = sprintf(self::CONIG_REGISTRY_KEY, $storeId);
        return $cacheId;
    }

    /**
     * Returns a separate cache id for a "null" store config.
     * A parent method could occasionally override cached store configuration with a parent website configuration.
     * Usually it occurs under a cache id 'store_1_payone_config_cache', where 1 is default store id.
     * @param int|null $storeId
     * @return string
     */
    public function getConfigCacheId($storeId = null)
    {
        if ($storeId === null) {
            $storeId = $this->getStoreId();
        }

        if ($storeId === null) {
            $storeId = self::NULL;
        }

        $cacheId = sprintf(self::CONFIG_CACHE_ID, $storeId);
        return $cacheId;
    }
}`

**Comment by fjbender** [comment]
@zaietsv Thanks, this sounds very reasonable for me!

@loswochos @Backenkoehler @tniebergall Can you weigh in on this?

**Comment by Backenkoehler** [comment]
@fjbender we are open to test this. 
@zaietsv do you have a repo for this?

**Comment by zaietsv** [comment]
@Backenkoehler Currently I haven't a separate module for this rewrite at this moment. Generally it is only a hot-fix for a Payone native module and mr. @fjbender will kindly provide a regular fix there. I can make a repository if you really need a separate module.

**Comment by Backenkoehler** [comment]
@zaietsv Ok, no problem. I guess there is a !== missing in your code (first if-clause)

**Comment by zaietsv** [comment]
@Backenkoehler I use exactly the same code at production and it works http://screenshots.collabstar.com/vza/a8f0047a.png

**Comment by fjbender** [comment]
@zaietsv Can you create a pull request with your changes? We'll then work it through with affected users and our QA team.

Thanks for your help on this one!

**Comment by Backenkoehler** [comment]
@zaietsv Without testing I would expect always haveing $storeId = 'NULL' because the first clause is true too but the variable gets overridden.

I have to test it, this are just my thoughts after looking at the code.

**Comment by mwr** [comment]
I have created a pull request adressing this issue. Details can be found in the pull-request.
https://github.com/PAYONE-GmbH/magento-1/pull/349

Regarding the Issue with: 
```
Call to a member function hasForcedState() on boolean in /app/code/core/Mage/Sales/Model/Order/Invoice.php 
```
This issue occurs in ``\Payone_Core_Model_Observer_TransactionStatus_InvoiceCreate::onPaid`` when the invoice is trying to be marked as paid but the invoice could not be found.
In those cases it is most likely that the Invoice should have been created during onAppointed but was not. The TxStatus APPOINTED for this order will then be in processing_status error.
We could double check if the invoice has actual data, but the root cause is probably with onAppointed not creating the invoice.
None the less we could detect the missing invoice and do something (log, mail, create the invoice)

**Comment by zaietsv** [comment]
> I have created a pull request adressing this issue. Details can be found in the pull-request.
> #349
> 
> Regarding the Issue with:
> 
> ```
> Call to a member function hasForcedState() on boolean in /app/code/core/Mage/Sales/Model/Order/Invoice.php 
> ```
> 
> This issue occurs in `\Payone_Core_Model_Observer_TransactionStatus_InvoiceCreate::onPaid` when the invoice is trying to be marked as paid but the invoice could not be found.
> In those cases it is most likely that the Invoice should have been created during onAppointed but was not. The TxStatus APPOINTED for this order will then be in processing_status error.
> We could double check if the invoice has actual data, but the root cause is probably with onAppointed not creating the invoice.
> None the less we could detect the missing invoice and do something (log, mail, create the invoice)

It's a pity but we have refused this module half-a-year ago. We hold it only for managing old orders in backend.

**Comment by hreinberger** [comment]
should be fixed by #349
